source/manuals/programming-languages/nodejs/style-guide.html.md.erb

```diff
@@ -0,0 +1,119 @@
+
+## Use strict mode
+
+[Strict mode] should be enabled for Node.js modules, as in the browser this prevents polluting the global scope (in
+Node.js this would pollute the `global` object, rather than the `window` object).
+
+## ES2015+
+
+Because with Node.js you control the version of the interpreter you're code is running on,
+you have access to many the latest features of JavaScript, such as:
+
+- [let] and [const]
+- [Arrow functions]
+- [Array.prototype] methods like `Array.map()`, `Array.find()` and `Array.values()`
+- The [spread operator]
+
+However use caution when using cutting edge JavaScript features and be sure to read the [documentation].
+
+A full list of what features are supported in each version of Node.js can be found at [Node Green].
+
+
+### let and const
+
+For Node.js code, be explicit about variable mutability using [const] and [let]:
+
+```
+const foo // GOOD: explicitly immutable variable declaration
+let bar // GOOD: expicitly mutable variable declaration
+var baz // BAD: not explicit about mutability of variable
+```
+
+This makes code easier to understand, as well as providing a feedback loop if a [const] variable is inadvertently reassigned.
+
+### Promises
+
+For Node.js applications, we use the native `Promise` implementation available in Node.js over alternatives like `q` or `bluebird`.
+This allows for a consistent Promise implementation between projects.
+
+Promise rejections should, as a general rule, reject with an instance of `Error` or one of it's subclasses. More on [Node.js Errors].
+
+```
+'use strict'
+
+// NPM Dependencies
+const request = require('request')
+
+function doSomethingAsync() {
+  return new Promise((resolve, reject) => {
+    request.get('https://www.example.com', (err, response, body) => {
+      if (err) {
+        reject(err)
+      } else {
+        resolve(body)
+      }
+    })
+  })
+}
+```
+
+
+## Organise your imports
+
+Import your dependencies at the top of each file, and group them into 'core', 'npm' and 'local' dependencies, like in the below example.
+This makes your imports clearer and allows others to  more easily understand and maintain cross dependencies of each file.
+
+```
+'use strict'
+
+// Core Dependencies
+const crypto = require('crypto')
+const fs = require('fs')
+const path = require('path')
+
+// NPM Dependencies
+const express = require('express')
+const lodash = require('lodash')
+const request = require('request')
+
+// Local Dependencies
+const config = require('../../config')
+const controller = require('./controller')
+const utilFunction = require('../../utils/util_function.js')
+```
+
+## Avoiding Unnecessary Function Declarations
+
+When building high volume applications in Node, it is important to be conscious of (and kind to) the V8 Engine's garbage collector.
+
+One area to be aware of, in order to maximise performance, is avoiding unnecessary function declarations within another function's scope.
```

Comment:
Would be worth adding a simple bad / good example of this in code to help aid understanding of the problem.

Reply:
I need to work on a generic example of this.

Reply:
@philandstuff I think this whole section needs clarifying. Basically we were using an anti-pattern on Pay which I'm trying to avoid being repeated here. @brid Is probably better placed to explain.

The pattern basically involved unnecessary functions were being declared, which were used to encapsulate a particular stage in the overall function's execution. While I can understand how this separated particular functionality into more human-readable stages, the inner functions were effectively being used in place of code comments, and had a negative performance impact.

A very-basic example is below, when the function is run, each of the inner functions has to be recreated in memory every time. When this pattern is used for writing multiple functions run along an express.js route, it leads to many unnecessary inner functions being created in memory every time the endpoint is hit. Overtime this can max out the memory so that the garbage collector just can't run, so the app falls over.

```
function doTheThing(someNumber) {
	const timesBy100 = () => {
		someNumber = someNumber * 100
	}

	const findOddness = () => {
		someNumber = someNumber % 2
	}

	const init = () => {
		timesBy100()
		findOddness()
	}

	init()

	return someNumber
}
```